src/Optimization/hiopKKTLinSys.cpp

```diff
@@ -2034,20 +2008,22 @@ bool hiopPrecondKKTOpr::combine_dir_to_build_vec(hiopVector& vec)
   size_type nx = dx_->get_size();
   size_type neq = dyc_->get_size();
   size_type nineq = dyd_->get_size();
-  assert(vec.get_local_size() == nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx+nx+nineq+nineq);
-
-  dx_->copyToStarting(   vec, 0);
-  dd_->copyToStarting(   vec, nx);
-  dyc_->copyToStarting(  vec, nx+nineq);
-  dyd_->copyToStarting(  vec, nx+nineq+neq);
-  dsxl_->copyToStarting( vec, nx+nineq+neq+nineq);
-  dsxu_->copyToStarting( vec, nx+nineq+neq+nineq+nx);
-  dsdl_->copyToStarting( vec, nx+nineq+neq+nineq+nx+nx);
-  dsdu_->copyToStarting( vec, nx+nineq+neq+nineq+nx+nx+nineq);
-  dzl_->copyToStarting(  vec, nx+nineq+neq+nineq+nx+nx+nineq+nineq);
-  dzu_->copyToStarting(  vec, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx);
-  dvl_->copyToStarting(  vec, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx+nx);
-  dvu_->copyToStarting(  vec, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx+nx+nineq);
+  assert(vec.get_size() == nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx+nx+nineq+nineq);
+
+  auto& dir_ir = dynamic_cast<hiopCompoundVector&>(vec);
+  dir_ir.getVector(0).copyFrom(*dx_);
+  dir_ir.getVector(1).copyFrom(*dd_);
+  dir_ir.getVector(2).copyFrom(*dyc_);
+  dir_ir.getVector(3).copyFrom(*dyd_);
+  dir_ir.getVector(4).copyFrom(*dsxl_);
+  dir_ir.getVector(5).copyFrom(*dsxu_);
+  dir_ir.getVector(6).copyFrom(*dsdl_);
+  dir_ir.getVector(7).copyFrom(*dsdu_);
+  dir_ir.getVector(8).copyFrom(*dzl_);
+  dir_ir.getVector(9).copyFrom(*dzu_);
+  dir_ir.getVector(10).copyFrom(*dvl_);
+  dir_ir.getVector(11).copyFrom(*dvu_);
```

Comment:
the order to build the compound vector is `hard-coded`. Should we use unordered_map to avoid this? @cnpetra

Reply:
It's still a draft PR. If we agree to this implementation, I will clean the code and add some comments.

Reply:
> the order to build the compound vector is `hard-coded`. Should we use unordered_map to avoid this? @cnpetra

I was thinking to use a `PrimalDualVector` class instead of the so-called compound. It may not covered all the use cases though...